python/ray/data/expressions.py

```diff
@@ -681,6 +694,29 @@ def structurally_equals(self, other: Any) -> bool:
         )
 
 
+@DeveloperAPI(stability="alpha")
+@dataclass(frozen=True, eq=False)
+class StarExpr(Expr):
+    """Expression that represents all columns from the input.
+
+    This is a special expression used in projections to indicate that
+    all existing columns should be preserved at this position in the output.
+    It's typically used internally by operations like with_column() and
+    rename_columns() to maintain existing columns.
+
+    Example:
+        When with_column("new_col", expr) is called, it creates:
+        Project(exprs=[star(), expr.alias("new_col")])
+
+        This means: keep all existing columns, then add/overwrite "new_col"
+    """
+
+    data_type: DataType = field(default_factory=lambda: DataType(object), init=False)
```

Comment:
Please leave a TODO to make StarExpr and Unresolved one down the line (unresolved expr won't have data-type since it's not resolved)